File: src/main/java/com/multicampus/gamesungcoding/a11ymarketserver/feature/address/service/AddressService.java
package com.multicampus.gamesungcoding.a11ymarketserver.feature.address.service;

import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.DataNotFoundException;
import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.UserNotFoundException;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.address.model.*;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.address.repository.AddressRepository;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.address.repository.DefaultAddressRepository;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.entity.Users;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class AddressService {

    private final AddressRepository addressRepository;
    private final DefaultAddressRepository defaultAddressRepository;
    private final UserRepository userRepository;

    // 배송지 목록 조회
    public List<AddressResponse> getAddressList(String userEmail) {
        return addressRepository.findByUserOrderByCreatedAtDesc(
                        getUserByEmail(userEmail)
                )
                .stream()
                .map(AddressResponse::fromEntity)
                .collect(Collectors.toList());
    }

    // 배송지 추가
    @Transactional
    public AddressResponse insertAddress(String userEmail, AddressRequest dto) {

        Addresses address = Addresses.builder()
                .user(getUserByEmail(userEmail))
                .addressInfo(
                        AddressInfo.builder()
                                .addressName(dto.addressName())
                                .receiverName(dto.receiverName())
                                .receiverPhone(dto.receiverPhone())
                                .receiverZipcode(dto.receiverZipcode())
                                .receiverAddr1(dto.receiverAddr1())
                                .receiverAddr2(dto.receiverAddr2())
                                .build()
                )
                .build();

        return AddressResponse.fromEntity(addressRepository.save(address));
    }

    // 배송지 수정
    @Transactional
    public AddressResponse updateAddress(String userEmail, String addressId, AddressRequest dto) {
        // 사용자 소유의 주소인지 확인
        addressRepository.findByAddressIdAndUser_UserId(
                        UUID.fromString(addressId),
                        getUserByEmail(userEmail).getUserId())
                .orElseThrow(() -> new DataNotFoundException("Address not found for user"));

        Addresses address = addressRepository.findById(UUID.fromString(addressId))
                .orElseThrow(() -> new DataNotFoundException("Address not found"));

        address.updateAddrInfo(
                AddressInfo.builder()
                        .addressName(dto.addressName())
                        .receiverName(dto.receiverName())
                        .receiverPhone(dto.receiverPhone())
                        .receiverZipcode(dto.receiverZipcode())
                        .receiverAddr1(dto.receiverAddr1())
                        .receiverAddr2(dto.receiverAddr2())
                        .build()
        );
        return AddressResponse.fromEntity(address);
    }

    // 배송지 삭제
    @Transactional
    public void deleteAddress(String userEmail, String addressId) {
        addressRepository.findByAddressIdAndUser_UserId(
                        UUID.fromString(addressId),
                        getUserByEmail(userEmail).getUserId()
                )
                .ifPresent(addressRepository::delete);
    }

    /**
     * Default_Address
     */

    // 기본 배송지 조회
    public AddressResponse getDefaultAddress(String userEmail) {
        return defaultAddressRepository.findByUser_UserId(getUserByEmail(userEmail).getUserId())
                .flatMap(defaultAddr -> addressRepository.findById(defaultAddr.getAddressId()))
                .map(AddressResponse::fromEntity)
                .orElseThrow(() -> new DataNotFoundException("Default address not found"));
    }

    // 기본 배송지 변경
    @Transactional
    public void setDefaultAddress(String userEmail, UUID addressId) {
        Users user = getUserByEmail(userEmail);

        // 사용자의 기본 배송지 정보가 존재하는지 확인 후 업데이트 또는 생성
        defaultAddressRepository.findById(user.getUserId())
                .ifPresentOrElse(
                        defaultAddress -> defaultAddress.changeDefaultAddress(addressId),
                        () -> defaultAddressRepository.save(DefaultAddress.builder()
                                .user(user)
                                .addressId(addressId)
                                .build())
                );
    }

    private Users getUserByEmail(String userEmail) {
        return userRepository.findByUserEmail(userEmail)
                .orElseThrow(() -> new UserNotFoundException("User not found"));
    }

}

File: src/main/kotlin/com/multicampus/gamesungcoding/a11ymarketserver/feature/address/service/AddressService.kt
package com.multicampus.gamesungcoding.a11ymarketserver.feature.address.service

import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.DataNotFoundException
import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.InvalidRequestException
import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.UserNotFoundException
import com.multicampus.gamesungcoding.a11ymarketserver.feature.address.dto.AddressRequest
import com.multicampus.gamesungcoding.a11ymarketserver.feature.address.dto.AddressResponse
import com.multicampus.gamesungcoding.a11ymarketserver.feature.address.entity.AddressInfo
import com.multicampus.gamesungcoding.a11ymarketserver.feature.address.entity.Addresses
import com.multicampus.gamesungcoding.a11ymarketserver.feature.address.mapper.toResponse
import com.multicampus.gamesungcoding.a11ymarketserver.feature.address.repository.AddressRepository
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.entity.Users
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.repository.UserRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
@Transactional(readOnly = true)
class AddressService(
    private val addressRepository: AddressRepository,
    private val userRepository: UserRepository
) {

    fun getAddressList(userId: UUID): List<AddressResponse> {
        return addressRepository.findByUserUserIdOrderByCreatedAtDesc(userId)
            .map { it.toResponse() }
    }

    @Transactional
    fun insertAddress(userId: UUID, req: AddressRequest): AddressResponse {
        var address = Addresses(
            user = getUserById(userId),
            address = AddressInfo(
                addressName = req.addressName,
                receiverName = req.receiverName,
                receiverPhone = req.receiverPhone,
                receiverZipcode = req.receiverZipcode,
                receiverAddr1 = req.receiverAddr1,
                receiverAddr2 = req.receiverAddr2
            ),
            isDefault = false
        )

        address = addressRepository.save(address)
        if (req.isDefault == true) {
            setDefaultAddressByAddressId(userId, requireNotNull(address.addressId))
        }

        return address.toResponse()
    }

    @Transactional
    fun updateAddress(userId: UUID, addressId: String, dto: AddressRequest): AddressResponse {
        val uuid = UUID.fromString(addressId)

        addressRepository.findByAddressIdAndUserUserId(uuid, userId)
            ?: throw DataNotFoundException("해당 사용자의 주소를 찾을 수 없습니다.")

        val address = addressRepository.findByIdOrNull(uuid)
            ?: throw DataNotFoundException("해당 주소를 찾을 수 없습니다.")

        if (dto.isDefault == true) {
            setDefaultAddressByAddressId(userId, uuid)
        }

        address.updateAddrInfo(
            AddressInfo(
                addressName = dto.addressName,
                receiverName = dto.receiverName,
                receiverPhone = dto.receiverPhone,
                receiverZipcode = dto.receiverZipcode,
                receiverAddr1 = dto.receiverAddr1,
                receiverAddr2 = dto.receiverAddr2
            )
        )

        return address.toResponse()
    }

    @Transactional
    fun deleteAddress(userId: UUID, addressId: String) {
        val address = addressRepository.findByAddressIdAndUserUserId(UUID.fromString(addressId), userId)
            ?: throw DataNotFoundException("해당 사용자의 주소를 찾을 수 없습니다.")

        if (address.isDefault) {
            throw InvalidRequestException("기본 배송지는 삭제할 수 없습니다.")
        }
        addressRepository.delete(address)
    }

    fun getDefaultAddress(userId: UUID): AddressResponse {
        val address = addressRepository.findByUserUserIdAndIsDefaultTrue(userId)
            ?: throw DataNotFoundException("기본 배송지가 설정되어 있지 않습니다.")
        return address.toResponse()
    }

    @Transactional
    fun setDefaultAddress(userId: UUID, addressId: UUID) {
        val defaultAddress = addressRepository.findByUserUserIdAndIsDefaultTrue(userId)

        if (defaultAddress != null) {
            if (defaultAddress.addressId == addressId) return
            defaultAddress.changeIsDefault(false)
        }

        setDefaultAddressByAddressId(userId, addressId)
    }

    private fun setDefaultAddressByAddressId(userId: UUID, addressId: UUID) {
        val address = addressRepository.findByIdOrNull(addressId)
            ?: throw DataNotFoundException("Address not found")

        if (address.user.userId != userId) {
            throw InvalidRequestException("Address does not belong to the user")
        }

        addressRepository.findByUserUserIdAndIsDefaultTrue(userId)?.let {
            if (it.addressId != addressId) it.changeIsDefault(false)
        }

        address.changeIsDefault(true)
    }

    private fun getUserById(userId: UUID): Users {
        return userRepository.findByIdOrNull(userId)
            ?: throw UserNotFoundException("사용자를 찾을 수 없습니다.")
    }
}